lib/Mojo/Util.pm

```diff
@@ -0,0 +1,1005 @@
+    }
+    $PATTERN{$pattern}->($str);
+  }
+  else { $str =~ s/([^A-Za-z0-9\-._~])/sprintf '%%%02X', ord $1/ge }
+
+  return $str;
+}
+
+sub url_unescape {
+  my $str = shift;
+  $str =~ s/%([0-9a-fA-F]{2})/chr hex $1/ge;
+  return $str;
+}
+
+sub xml_escape {
+  return $_[0] if ref $_[0] && ref $_[0] eq 'Mojo::ByteStream';
+  my $str = shift // '';
+  $str =~ s/([&<>"'])/$XML{$1}/ge;
+  return $str;
+}
+
+sub xor_encode {
+  my ($input, $key) = @_;
+
+  # Encode with variable key length
+  my $len    = length $key;
+  my $buffer = my $output = '';
+  $output .= $buffer ^ $key while length($buffer = substr($input, 0, $len, '')) == $len;
+  return $output .= $buffer ^ substr($key, 0, length $buffer, '');
+}
+
+sub _adapt {
+  my ($delta, $numpoints, $firsttime) = @_;
+  use integer;
+
+  $delta = $firsttime ? $delta / PC_DAMP : $delta / 2;
+  $delta += $delta / $numpoints;
+  my $k = 0;
+  while ($delta > ((PC_BASE - PC_TMIN) * PC_TMAX) / 2) {
+    $delta /= PC_BASE - PC_TMIN;
+    $k     += PC_BASE;
+  }
+
+  return $k + (((PC_BASE - PC_TMIN + 1) * $delta) / ($delta + PC_SKEW));
+}
+
+sub _encoding { $ENCODING{$_[0]} //= find_encoding($_[0]) // croak "Unknown encoding '$_[0]'" }
+
+sub _entity {
+  my ($point, $name, $attr) = @_;
+
+  # Code point
+  return chr($point !~ /^x/ ? $point : hex $point) unless defined $name;
+
+  # Named character reference
+  my $rest = my $last = '';
+  while (length $name) {
+    return $ENTITIES{$name} . reverse $rest
+      if exists $ENTITIES{$name} && (!$attr || $name =~ /;$/ || $last !~ /[A-Za-z0-9=]/);
+    $rest .= $last = chop $name;
+  }
+  return '&' . reverse $rest;
+}
+
+sub _header {
+  my ($str, $cookie) = @_;
+
+  my (@tree, @part);
+  while ($str =~ /\G[,;\s]*([^=;, ]+)\s*/gc) {
+    push @part, $1, undef;
+    my $expires = $cookie && @part > 2 && lc $1 eq 'expires';
+
+    # Special "expires" value
+    if ($expires && $str =~ /\G=\s*$EXPIRES_RE/gco) { $part[-1] = $1 }
+
+    # Quoted value
+    elsif ($str =~ /$QUOTED_VALUE_RE/gco) { $part[-1] = unquote $1 }
+
+    # Unquoted value
+    elsif ($str =~ /$UNQUOTED_VALUE_RE/gco) { $part[-1] = $1 }
+
+    # Separator
+    next unless $str =~ /\G[;\s]*,\s*/gc;
+    push @tree, [@part];
+    @part = ();
+  }
+
+  # Take care of final part
+  return [@part ? (@tree, \@part) : @tree];
+}
+
+sub _html {
+  my ($str, $attr) = @_;
+  $str =~ s/$ENTITY_RE/_entity($1, $2, $attr)/geo;
+  return $str;
+}
+
+sub _options {
+
+  # Hash or name (one)
+  return ref $_[0] eq 'HASH' ? (undef, %{shift()}) : @_ if @_ == 1;
+
+  # Name and values (odd)
+  return shift, @_ if @_ % 2;
+
+  # Name and hash or just values (even)
+  return ref $_[1] eq 'HASH' ? (shift, %{shift()}) : (undef, @_);
+}
+
+# This may break in the future, but is worth it for performance
+sub _readable { !!(IO::Poll::_poll(@_[0, 1], my $m = POLLIN | POLLPRI) > 0) }
+
+sub _round { $_[0] < 10 ? int($_[0] * 10 + 0.5) / 10 : int($_[0] + 0.5) }
+
+sub _stash {
```

Comment:
This seems like a really nice function, and I've copied it into my own modules sometimes.  Is there a reason not to make this a public function?